### DISCUSSION TITLE
Any way to get finsihed jobs?

### DISCUSSION THREAD

**Discussion by hldup** [apalis-dev/apalis#340] [Q&A] [answered]
Hey, I have an axum server that deploys jobs to different kinds of microservices. All it basically does is send jobs to the microservices and wait do other stuff in the meanwhile. Then once it has finished I want it to check up on the job I have dispatched to the microservice.
Is there a function or any way to query the RedisStorage to check if a job has finished?

**Comment by geofmureithi** [comment]
All storage have a `fetch_by_id` method. It should give you the updated
version of a job.

**Reply by hldup** [reply]
Thanks for the reply! But i still have some problems

**Comment by hldup** [comment]
I have made a basic wrapper function that executes a trait function that is implemented to a struct that implements apalis::prelude::Job. 
After the job has finished I still get the wrong data back from redis. What could I be doing wrong?

*main.rs*
```rust
     Monitor::<TokioExecutor>::new()
            .register_with_count(4, {
                WorkerBuilder::new("edit-runner")
                    .layer(TraceLayer::new())
                    .with_storage(storage.clone())
                    .data(state.clone())
                    .build_fn(jobhandler::run_job::<EditorJob, State>)
            })
            .run()
            .await
            .unwrap();
```

*jobhandler::run_job()*
```rust
pub async fn run_job<T, S>(mut job: T, state: Data<S>)
where
    T: Serialize + DeserializeOwned + Send + 'static + Unpin + Job + Sync + JobRunner<T, S>,
    S: StateHandler,
{
    match job.run(state).await {
        Ok(_) => {
            log::debug!("Job has finished ok");
            job.set_status(JobStatus::Ok)
        }
        Err(error) => {
            log::error!("job has failed");
            job.set_status(JobStatus::Error("failed".into()))
        }
    };
}
```
*job.rs*
```rust
impl JobRunner<EditorJob, crate::State> for EditorJob {
    async fn run(&mut self, state: Data<crate::State>) -> Result<(), jobhandler::JobError> {
        thread::sleep(Duration::from_secs(35));
        Ok(())
    }
    fn set_status(&mut self, status: jobhandler::JobStatus) {
        self.set_status(status);
    }
    fn get_status(&self) -> jobhandler::JobStatus {
        self.status().to_owned()
    }
}
```

**Reply by hldup** [reply]
![20240702_0408](https://github.com/geofmureithi/apalis/assets/73071529/94c8dfa4-7f03-4fe3-b4d3-6021c7b8676d)
Tho after it has run, it is recorded in the editor::job::done table.

**Comment by geofmureithi** [comment]
Does fetch_with_id + update work for you?
https://docs.rs/apalis-redis/0.6.0-alpha.0/src/apalis_redis/storage.rs.html#657-671

**Reply by hldup** [reply]
Let me try, I'll migrate to 0.6.0 really quick

**Reply by hldup** [reply]
No, I don't seem to have a function called `update` on Request<T>. 
![20240702_0910](https://github.com/geofmureithi/apalis/assets/73071529/eb72b7b1-fb11-4384-a4f3-407bd57e00e5).

Am I supposed to update the Job in the function I pass to `build_fn`? Because of yes I need to pass a state to my function that has access to the redis connection.

This is my current function that handles the jobs
```rust
pub async fn run_job<T, S>(
    mut job: T,
    state: Data<S>,
    worker_id: WorkerId,
) -> Result<(), apalis::prelude::Error>
where
    T: Serialize + DeserializeOwned + Send + 'static + Unpin + Sync + JobRunner<T, S> + Clone,
    S: Clone,
{
    log::debug!("▶ Started a job with worker: {}", worker_id);
    match job.run(state.clone()).await {
        Ok(_) => {
            log::debug!("Job has finished ok {}", worker_id.to_string());
            job.set_status(JobStatus::Ok);
        }
        Err(error) => {
            log::error!("{}", error);
            job.set_status(JobStatus::Error("Asd".into()))
        }
    };
    Ok(())
}
```

**Reply by geofmureithi** [reply]
update is a method of the storage not the request

**Reply by hldup** [reply]
To be honest I am not familiar with `tower` so I have no idea what I am really doing but this has been blocking my progress on my project. 
I tried the following
```rust
match conn.update(Request::new(job.clone())).await {
      Ok(_) => {
          log::info!("job updated in redis")
      }
      Err(error) => {
          log::error!("Failed to update job in redis: {}", error)
      }
  }
```

Which returns
```
 ERROR job: jobhandler: Failed to update job in redis: Missing TaskId- IoError    
```

**Reply by hldup** [reply]
For a temporary fix  I just wrote my own code to fetch the id of the job from the sorted set to determine if its done or not.

**Reply by hldup** [reply]
Is the job suppossed to automatically update after its done? @geofmureithi

**Reply by geofmureithi** [reply]
I am currently working on saving results. This issue #182 has some more context.

**Reply by hldup** [reply]
I migrated to the newest release as of now, I see that you implemented saving of results. (Great work, thank you!) 
But when I fetch from the storage I get the initial state of the job not the result. Is there supposed to be a function that fetches results?
Received data from: jobhandler::node::Job:data
Instead of jobhandler::node::Job:data::result

**Reply by geofmureithi** [reply, marked as answer]
Its still WIP, I need to incorporate the result part. You could pull it manually if needed.

**Reply by hldup** [reply]
Thanks, I'll write my own method to fetch from redis till than.

**Reply by hldup** [reply]
I wrote a little code that fetches from the hashset and decodes the response but apalis returns a `Encountered an error during streaming`, any ideas what I could be doing wrong during decoding?
```rust
    let job: Value = match conn
        .hget(
            format!("{}:data::result", std::any::type_name::<T>()),
            id.clone().to_string(),
        )
        .await
    {
        Err(error) => {
            log::warn!(
                "Job for id: {} does not exist {}",
                id.to_string(),
                error
            );
            return Err(StatusCode::NOT_FOUND);
        }
        Ok(job) => job,
    };

    match job {
        Value::Nil => return Err(StatusCode::NOT_FOUND),
        Value::Data(job) => {
            match storage.get_codec().decode(&job) {
                Err(error) => {
                    log::error!("Failed to decode result from redis: {}", error);
                    return Err(StatusCode::INTERNAL_SERVER_ERROR);
                }
                Ok(job) => return Ok(job.job().to_owned().into()),
            };
        }
        _ => return Err(StatusCode::INTERNAL_SERVER_ERROR),
    };
    ```

**Reply by hldup** [reply]
Ah darn, I overlooked the data itself, the result is not saved as json. That's the main reason I get an error. stupid me

**Comment by kdesjard** [comment]
Is there an easy way to decode last_error back to Result<T,String> ?

**Reply by geofmureithi** [reply]
On Redis not yet, but on sql backends its stored as json for `Result<String, String>`, I need to improve this and am open to ideas.

**Reply by hldup** [reply]
AFAIK Only PostgreSQL has an explicit json format for columns so I think storing the result and error as a Json string is the best option

**Comment by kdesjard** [comment]
I tried the following and it worked for me:

In layers.rs:
```
impl<SV, A, T> Service<Request<T>> for AckService<SV, A, T>
...
```
if I convert the ok result to a serde_json::Value,instead of format!({ok:?}):
eg:
```
.map(|ok| serde_json::to_value(ok).unwrap_or_default())
```
and then in my own code, check last_error for a an option value, json decode the value back into a Result<MyType,String> and it works.

**Reply by hldup** [reply]
Yeah that's what I was thinking of too, 
chaging [this line in the code](https://github.com/geofmureithi/apalis/blob/cc58602b7e284cf258ddcb78c01e5ea988109b0e/packages/apalis-redis/src/storage.rs#L550) to not store the result as a string, instead just use serde_json to turn it into a value and then turn it into a string and save it in redis.

**Reply by kdesjard** [reply]
Ideally, the Error from the executed function would be put in the Result that is stored in the DB. I have my own Error type and what gets stored now is "FailedError: Unknown".  Maybe I'm missing a trait or something.

**Comment by kdesjard** [comment]
Here is a changeset that is working for me (sqlite only right now) and currently has a couple of unwraps:

https://github.com/kdesjard/apalis/commit/ecaac411232c02b22048f22e9f9ca9160019f48f

**Reply by geofmureithi** [reply]
@kdesjard Here is a PR that incorporates what you are trying to do.
https://github.com/geofmureithi/apalis/pull/383
